docs/devel/coding-conventions.md

```diff
@@ -30,12 +30,48 @@ Documentation for other releases can be found at
 <!-- END STRIP_FOR_RELEASE -->
 
 <!-- END MUNGE: UNVERSIONED_WARNING -->
-Coding style advice for contributors
+Code conventions
   - Bash
     - https://google-styleguide.googlecode.com/svn/trunk/shell.xml
   - Go
-    - https://github.com/golang/go/wiki/CodeReviewComments
-    - https://gist.github.com/lavalamp/4bd23295a9f32706a48f
+    - Ensure your code passes the [presubmit checks](development.md#hooks)
+    - [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
+    - [Effective Go](https://golang.org/doc/effective_go.html)
+    - Command-line flags should use dashes, not underscores
+    - API conventions
+      - [API changes](api_changes.md)
+      - [API conventions](api-conventions.md)
+    - [Kubectl conventions](kubectl-conventions.md)
+    - [Logging conventions](logging.md)
+
+Testing conventions
+  - All new packages and most new significant functionality must come with unit tests
+  - Table-driven tests are preferred for testing multiple scenarios/inputs; for example, see [TestNamespaceAuthorization](../../test/integration/auth_test.go)
+  - Significant features should come with integration (test/integration) and/or end-to-end (test/e2e) tests
+    - Including new kubectl commands and major features of existing commands
+
+Directory and file conventions
+    - Avoid package sprawl (see [#4851](http://issues.k8s.io/4851) for discussion)
+      - General utilities belong in new package subdirectories of pkg/util
```

Comment:
Avoid general utility packages in general - packages called "util" are suspect.  Instead, derive a name that describes your desired function.  For example, the utility functions dealing with waiting for operations are in the `wait` package (under `util`) and include functionality like `Poll`.  So the full name is `wait.Poll`